sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/schema/common/SchemaManager.java

```diff
@@ -0,0 +1,213 @@
+            });
+  }
+
+  // gets a keyspace by provided name, no transformations applied
+  private Uni<Schema.CqlKeyspaceDescribe> getKeyspaceInternal(
+      StargateBridge bridge, String keyspaceName) {
+    Optional<String> tenantId = requestInfo.getTenantId();
+
+    // check if cached, if so we need to revalidate hash
+    CompositeCacheKey cacheKey = new CompositeCacheKey(keyspaceName, tenantId);
+    boolean cached = keyspaceCache.as(CaffeineCache.class).keySet().contains(cacheKey);
+
+    // get keyspace from cache
+    return fetchKeyspace(keyspaceName, tenantId, bridge)
+
+        // check hash still matches
+        .flatMap(
+            keyspace -> {
+              // if it was not cached before, we can simply return
+              // no need to check
+              if (!cached) {
+                return Uni.createFrom().item(keyspace);
+              }
+
+              Schema.DescribeKeyspaceQuery request =
+                  Schema.DescribeKeyspaceQuery.newBuilder()
+                      .setKeyspaceName(keyspaceName)
+                      .setHash(keyspace.getHash())
+                      .build();
+
+              // call bridge
+              return bridge
+                  .describeKeyspace(request)
+                  .onItem()
+                  .transformToUni(
+                      updatedKeyspace -> {
+                        // if we have updated keyspace cache and return
+                        // otherwise return what we had in the cache already
+                        if (null != updatedKeyspace && updatedKeyspace.hasCqlKeyspace()) {
+                          invalidateKeyspace(keyspaceName, tenantId);
+                          return cacheKeyspace(keyspaceName, tenantId, updatedKeyspace);
+                        } else {
+                          return Uni.createFrom().item(keyspace);
+                        }
+                      });
+            })
+
+        // in case of failure, check if status is not found
+        // and invalidate the cache
+        .onFailure()
+        .recoverWithUni(
+            t -> {
+              if (t instanceof StatusRuntimeException sre) {
+                if (Objects.equals(sre.getStatus().getCode(), Status.Code.NOT_FOUND)) {
+                  invalidateKeyspace(keyspaceName, tenantId);
+                  return Uni.createFrom().nullItem();
+                }
+              }
+
+              return Uni.createFrom().failure(t);
+            });
+  }
+
+  // fetches the keyspace from the bridge and caches it
+  @CacheResult(cacheName = "keyspace-cache")
+  protected Uni<Schema.CqlKeyspaceDescribe> fetchKeyspace(
+      @CacheKey String keyspaceName, @CacheKey Optional<String> tenantId, StargateBridge bridge) {
+    // create request
+    Schema.DescribeKeyspaceQuery request =
+        Schema.DescribeKeyspaceQuery.newBuilder().setKeyspaceName(keyspaceName).build();
+
+    // memoize
+    return bridge.describeKeyspace(request).memoize().indefinitely();
```

Comment:
Should this be indefinite? Is it memoizing the result? What happens if someone drops a keyspace and then re-creates it with different replication or other options?

Reply:
This is a memoize of the uni that's being cached, the only purpose is not to re-execute the request on the bridge once the item is pulled in the cache.. With the reactive frameworks in case of a double subscription to the same producer, the whole chain is re-executed..